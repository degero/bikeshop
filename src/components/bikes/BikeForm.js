import React from "react";
import TextInput from "../common/TextInput";

const BikeForm = ({ bike, onSave, onChange, errors = {}, saving = false }) => (
  <div className="row content">
    <div className="col-12 col-md-4">
      <h2>{bike.id ? "Edit" : "Create"} Bike</h2>
      {errors.onSave && (
        <div className="alert alert-danger" role="alert">
          {errors.onSave}
        </div>
      )}
      <form onSubmit={onSave} autoComplete="off">
        <TextInput
          name="manufacturer"
          label="Manufacturer:"
          onChange={onChange}
          value={bike.manufacturer}
          error={errors.manufacturer}
        ></TextInput>
        <TextInput
          name="model"
          label="Model:"
          onChange={onChange}
          value={bike.model}
          error={errors.model}
        ></TextInput>
        <TextInput
          name="price"
          label="Price ($USD):"
          onChange={onChange}
          value={bike.price}
          error={errors.price}
        ></TextInput>
        <input
          className="btn btn-primary"
          type="submit"
          value={saving ? "Saving" : "Save"}
        />
      </form>
    </div>
  </div>
);

export default BikeForm;
