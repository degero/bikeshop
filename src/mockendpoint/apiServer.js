/* json-server for dev use, see .env.development api endpoint setting */
const jsonServer = require("json-server");
const server = jsonServer.create();
const path = require("path");
const router = jsonServer.router(path.join(__dirname, "db.json"));

const middleware = jsonServer.defaults();
server.use(middleware);

server.use(jsonServer.bodyParser);

// Slow call down 1500ms
server.use(function (req, res, next) {
  setTimeout(next, 1500);
});

server.use((req, res, next) => {
  // set a create date on post
  if (req.method === "POST") {
    req.body.createdAt = Date.now();
  }
  // TODO add updated date on PUT
  next();
});

server.post("/bikes/", function (req, res, next) {
  const error = validateBike(req.body);
  if (error) {
    res.status(400).send(error);
  } else {
    req.body.slug = createSlug(req.body.manufacturer + " " + req.body.model); // Generate a slug for new bike.
    next();
  }
});

server.use(router);

// App runs on 3000 and api on 3001
const port = 3001;
server.listen(port, () => {
  console.log(`JSON Server is running on port ${port}`);
});

function createSlug(value) {
  return value
    .replace(/[^a-z0-9_]+/gi, "-")
    .replace(/^-|-$/g, "")
    .toLowerCase();
}

// Server side validation of bike
function validateBike(bike) {
  if (!bike.manufacturer) return "Manufacturer is required.";
  if (!bike.model) return "Model is required.";
  if (!bike.price) return "Price is required.";
  return "";
}
