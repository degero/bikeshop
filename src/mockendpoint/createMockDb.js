/* eslint-disable no-console */
const fs = require("fs");
const path = require("path");
const mockData = require("./mockData");

const { bikes } = mockData;
const data = JSON.stringify({ bikes });
const filepath = path.join(__dirname, "db.json");

fs.writeFile(filepath, data, function (err) {
  err ? console.log(err) : console.log("Mock DB db.json created.");
});
