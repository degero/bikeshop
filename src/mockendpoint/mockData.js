const bikes = [
  {
    id: 1,
    manufacturer: "Santa Cruz",
    model: "Bronson",
    price: 6999,
    slug: "santa-cruz-bronson",
  },
];

const manufacturers = [
  {
    id: 1,
    name: "Santa Cruz",
    slug: "santa-cruz"
  },
];

module.exports = {
  bikes,
  manufacturers
};
