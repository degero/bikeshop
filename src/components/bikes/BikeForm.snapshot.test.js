import React from "react";
import renderer from "react-test-renderer";
import BikeForm from "./BikeForm";
import { bikes, manufacturers } from "../../mockendpoint/mockData";

describe("Bikeform component - snapshots", () => {
  it("sets submit button label 'Saving...' when saving is true", () => {
    // ARRANGE / ACT
    const tree = renderer.create(
      <BikeForm
        bike={bikes[0]}
        manufacturers={manufacturers}
        onSave={jest.fn()}
        onChange={jest.fn()}
        saving={true}
      />
    );
    // ASSERT
    expect(tree).toMatchSnapshot();
  });

  it("sets submit button label 'Save' when saving is false", () => {
    // ARRANGE / ACT
    const tree = renderer.create(
      <BikeForm
        bike={bikes[0]}
        manufacturers={manufacturers}
        onSave={jest.fn()}
        onChange={jest.fn()}
        saving={false}
      />
    );
    // ASSERT
    expect(tree).toMatchSnapshot();
  });
});
