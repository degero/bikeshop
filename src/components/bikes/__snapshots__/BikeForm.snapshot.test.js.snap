// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bikeform component - snapshots sets submit button label 'Save' when saving is false 1`] = `
<div
  className="row content"
>
  <div
    className="col-12 col-md-4"
  >
    <h2>
      Edit
       Bike
    </h2>
    <form
      autoComplete="off"
      onSubmit={[MockFunction]}
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="manufacturerId"
        >
          Manufacturer:
        </label>
        <select
          className="form-control"
          name="manufacturerId"
          onChange={[Function]}
          value="1"
        >
          <option
            value=""
          >
            Please select
          </option>
          <option
            value={1}
          >
            Santa Cruz
          </option>
          <option
            value={2}
          >
            Yeti
          </option>
        </select>
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="model"
        >
          Model:
        </label>
        <input
          className="form-control"
          name="model"
          onChange={[MockFunction]}
          type="text"
          value="Bronson"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="price"
        >
          Price ($USD):
        </label>
        <input
          className="form-control"
          name="price"
          onChange={[MockFunction]}
          type="text"
          value={6999}
        />
      </div>
      <input
        className="btn btn-primary"
        type="submit"
        value="Save"
      />
    </form>
  </div>
</div>
`;

exports[`Bikeform component - snapshots sets submit button label 'Saving...' when saving is true 1`] = `
<div
  className="row content"
>
  <div
    className="col-12 col-md-4"
  >
    <h2>
      Edit
       Bike
    </h2>
    <form
      autoComplete="off"
      onSubmit={[MockFunction]}
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="manufacturerId"
        >
          Manufacturer:
        </label>
        <select
          className="form-control"
          name="manufacturerId"
          onChange={[Function]}
          value="1"
        >
          <option
            value=""
          >
            Please select
          </option>
          <option
            value={1}
          >
            Santa Cruz
          </option>
          <option
            value={2}
          >
            Yeti
          </option>
        </select>
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="model"
        >
          Model:
        </label>
        <input
          className="form-control"
          name="model"
          onChange={[MockFunction]}
          type="text"
          value="Bronson"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="price"
        >
          Price ($USD):
        </label>
        <input
          className="form-control"
          name="price"
          onChange={[MockFunction]}
          type="text"
          value={6999}
        />
      </div>
      <input
        className="btn btn-primary"
        type="submit"
        value="Saving"
      />
    </form>
  </div>
</div>
`;
